File: backend/app/routers/conflicts.py
from datetime import datetime
from fastapi import APIRouter, HTTPException, status, Depends
from bson import ObjectId
from app.database import get_database
from app.models.conflict import ConflictCreate, Conflict
from app.utils.auth import get_current_user_id
from app.utils.trip_permissions import check_trip_access

router = APIRouter(prefix="/trips/{trip_id}/conflicts", tags=["conflicts"])


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_conflict(
    trip_id: str,
    conflict_data: ConflictCreate,
    user_id: str = Depends(get_current_user_id)
):
    """Create a conflict (typically by agent)"""
    await check_trip_access(trip_id, user_id)
    
    db = get_database()
    
    # Verify message exists and belongs to trip
    message = await db.messages.find_one({
        "_id": ObjectId(conflict_data.message_id),
        "tripId": ObjectId(trip_id)
    })
    if not message:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Message not found")
    
    now = datetime.utcnow()
    conflict_doc = {
        "tripId": ObjectId(trip_id),
        "messageId": ObjectId(conflict_data.message_id),
        "options": [opt.dict() for opt in conflict_data.options],
        "createdAt": now,
    }
    
    result = await db.conflicts.insert_one(conflict_doc)
    conflict_doc["_id"] = result.inserted_id
    
    # Update message with conflict reference
    await db.messages.update_one(
        {"_id": ObjectId(conflict_data.message_id)},
        {"$set": {"conflictId": conflict_doc["_id"]}}
    )
    
    return {
        "id": str(conflict_doc["_id"]),
        "trip_id": trip_id,
        "message_id": conflict_data.message_id,
        "options": conflict_doc["options"],
        "created_at": conflict_doc.get("createdAt"),
    }


@router.post("/{conflict_id}/vote")
async def vote_on_conflict(
    trip_id: str,
    conflict_id: str,
    vote_data: dict,
    user_id: str = Depends(get_current_user_id)
):
    option_key = vote_data.get("option_key") or vote_data.get("key")
    """Vote on a conflict option"""
    await check_trip_access(trip_id, user_id)
    
    db = get_database()
    
    conflict = await db.conflicts.find_one({
        "_id": ObjectId(conflict_id),
        "tripId": ObjectId(trip_id)
    })
    if not conflict:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Conflict not found")
    
    # Check if option exists
    option_index = next(
        (i for i, opt in enumerate(conflict["options"]) if opt["key"] == option_key),
        None
    )
    if option_index is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Option not found")
    
    user_obj_id = ObjectId(user_id)
    
    # Check if user already voted (remove old vote)
    for opt in conflict["options"]:
        opt["votes"] = [v for v in opt.get("votes", []) if v.get("userId") != user_obj_id]
    
    # Add new vote
    conflict["options"][option_index]["votes"].append({
        "userId": user_obj_id,
        "at": datetime.utcnow()
    })
    
    await db.conflicts.update_one(
        {"_id": ObjectId(conflict_id)},
        {"$set": {"options": conflict["options"]}}
    )
    
    return {"message": "Vote recorded"}


@router.get("/{conflict_id}")
async def get_conflict(
    trip_id: str,
    conflict_id: str,
    user_id: str = Depends(get_current_user_id)
):
    """Get a specific conflict"""
    await check_trip_access(trip_id, user_id)
    
    db = get_database()
    conflict = await db.conflicts.find_one({
        "_id": ObjectId(conflict_id),
        "tripId": ObjectId(trip_id)
    })
    
    if not conflict:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Conflict not found")
    
    return {
        "id": str(conflict["_id"]),
        "trip_id": trip_id,
        "message_id": str(conflict["messageId"]),
        "options": conflict["options"],
        "created_at": conflict.get("createdAt"),
    }

File: backend/src/routers/conflicts.ts
import { Router, Response } from 'express';
import { ObjectId } from 'mongodb';
import { getDatabase } from '../database.js';
import { ConflictCreateSchema } from '../models/conflict.js';
import { getCurrentUserId, AuthRequest } from '../utils/auth.js';
import { checkTripAccess } from '../utils/trip_permissions.js';
import { sseService } from '../services/sse.js';

const router = Router({ mergeParams: true });

router.post('', getCurrentUserId, async (req: AuthRequest, res: Response) => {
  try {
    const { trip_id } = req.params;
    const validated = ConflictCreateSchema.parse(req.body);
    
    await checkTripAccess(trip_id, req.userId!);
    
    const db = getDatabase();
    
    // Verify message exists and belongs to trip
    const message = await db.collection('messages').findOne({
      _id: new ObjectId(validated.message_id),
      tripId: new ObjectId(trip_id),
    });
    
    if (!message) {
      res.status(404).json({ detail: 'Message not found' });
      return;
    }
    
    const now = new Date();
    const conflictDoc = {
      tripId: new ObjectId(trip_id),
      messageId: new ObjectId(validated.message_id),
      options: validated.options,
      createdAt: now,
    };
    
    const result = await db.collection('conflicts').insertOne(conflictDoc);
    
    // Update message with conflict reference
    await db.collection('messages').updateOne(
      { _id: new ObjectId(validated.message_id) },
      { $set: { conflictId: result.insertedId } }
    );
    
    const conflictResponse = {
      id: result.insertedId.toString(),
      trip_id,
      message_id: validated.message_id,
      options: conflictDoc.options,
      created_at: conflictDoc.createdAt,
    };
    
    // Broadcast conflict via SSE
    sseService.broadcastConflict(trip_id, conflictResponse);
    
    res.status(201).json(conflictResponse);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ detail: 'Validation error', errors: error.errors });
      return;
    }
    if (error.message === 'Trip not found' || error.message === 'Access denied') {
      res.status(error.message === 'Trip not found' ? 404 : 403).json({ detail: error.message });
    } else {
      res.status(500).json({ detail: error.message });
    }
  }
});

router.post('/:conflict_id/vote', getCurrentUserId, async (req: AuthRequest, res: Response) => {
  try {
    const { trip_id, conflict_id } = req.params;
    const { option_key, key } = req.body;
    const optionKey = option_key || key;
    
    await checkTripAccess(trip_id, req.userId!);
    
    const db = getDatabase();
    const conflict = await db.collection('conflicts').findOne({
      _id: new ObjectId(conflict_id),
      tripId: new ObjectId(trip_id),
    });
    
    if (!conflict) {
      res.status(404).json({ detail: 'Conflict not found' });
      return;
    }
    
    const optionIndex = conflict.options.findIndex((opt: any) => opt.key === optionKey);
    if (optionIndex === -1) {
      res.status(404).json({ detail: 'Option not found' });
      return;
    }
    
    const userObjId = new ObjectId(req.userId!);
    
    // Remove old vote from all options
    for (const opt of conflict.options) {
      opt.votes = (opt.votes || []).filter((v: any) => !v.userId.equals(userObjId));
    }
    
    // Add new vote
    if (!conflict.options[optionIndex].votes) {
      conflict.options[optionIndex].votes = [];
    }
    conflict.options[optionIndex].votes.push({
      userId: userObjId,
      at: new Date(),
    });
    
    await db.collection('conflicts').updateOne(
      { _id: new ObjectId(conflict_id) },
      { $set: { options: conflict.options } }
    );
    
    // Broadcast vote update via SSE
    const voteCount = conflict.options[optionIndex].votes.length;
    sseService.broadcastVote(trip_id, conflict_id, optionKey, voteCount);
    
    res.json({ message: 'Vote recorded' });
  } catch (error: any) {
    if (error.message === 'Trip not found' || error.message === 'Access denied') {
      res.status(error.message === 'Trip not found' ? 404 : 403).json({ detail: error.message });
    } else {
      res.status(500).json({ detail: error.message });
    }
  }
});

router.get('/:conflict_id', getCurrentUserId, async (req: AuthRequest, res: Response) => {
  try {
    const { trip_id, conflict_id } = req.params;
    await checkTripAccess(trip_id, req.userId!);
    
    const db = getDatabase();
    const conflict = await db.collection('conflicts').findOne({
      _id: new ObjectId(conflict_id),
      tripId: new ObjectId(trip_id),
    });
    
    if (!conflict) {
      res.status(404).json({ detail: 'Conflict not found' });
      return;
    }
    
    res.json({
      id: conflict._id.toString(),
      trip_id,
      message_id: conflict.messageId.toString(),
      options: conflict.options,
      created_at: conflict.createdAt,
    });
  } catch (error: any) {
    if (error.message === 'Trip not found' || error.message === 'Access denied') {
      res.status(error.message === 'Trip not found' ? 404 : 403).json({ detail: error.message });
    } else {
      res.status(500).json({ detail: error.message });
    }
  }
});

export default router;
